runtime/sema/check_reference_expression.go

```diff
@@ -47,35 +54,24 @@ func (checker *Checker) VisitReferenceExpression(referenceExpression *ast.Refere
 			)
 		} else {
 			targetType = referenceType.Type
+			if optOk {
+				targetType = &OptionalType{Type: targetType}
+			}
 		}
 	}
+	returnType = referenceType
 
 	// Type-check the referenced expression
 
 	referencedExpression := referenceExpression.Expression
 
-	// If the referenced expression is an index expression, it might be into storage
-
-	indexExpression, isIndexExpression := referencedExpression.(*ast.IndexExpression)
-	if isIndexExpression {
-		// The referenced expression will evaluate to an optional type if it is indexing:
-		// the result of the access is an optional.
-		//
-		// Hence expect an optional.
-
-		expectedType := wrapWithOptionalIfNotNil(targetType)
-
-		_, referencedType = checker.visitExpression(indexExpression, expectedType)
-
-		// Unwrap the optional one level, but not infinitely
-
-		if optionalReferencedType, ok := referencedType.(*OptionalType); ok {
-			referencedType = optionalReferencedType.Type
-		}
+	_, referencedType = checker.visitExpression(referencedExpression, targetType)
 
-	} else {
-		// If the referenced expression is not an index expression, check it normally
-		_, referencedType = checker.visitExpression(referencedExpression, targetType)
+	// Unwrap the optional one level, but not infinitely
+	if optionalReferencedType, ok := referencedType.(*OptionalType); ok {
+		// if referenced type is optional, require that target type is also optional
+		referencedType = optionalReferencedType.Type
+		returnType = &OptionalType{Type: referenceType}
 	}
 
 	if _, ok := referencedType.(*OptionalType); ok {
```

Comment:
This if-statement looks like dead code now: It is the same code as above (`referencedType.(*OptionalType)`), and cannot occur anymore, as `referencedType` is unwrapped and re-assigned above.

I think this statement can be removed, and the error type `OptionalTypeReferenceError` too.